src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -370,13 +562,339 @@ private int LookupAggregatorStore(KeyValuePair<string, object?>[] tagKeysAndValu
+
+                givenTags = new Tags(givenTagKeysAndValues);
+
+                Debug.Assert(this.availableMetricPoints != null, "this.availableMetricPoints was null");
+
+                lock (this.tagsToMetricPointIndexDictionaryDelta)
+                {
+                    // check again after acquiring lock.
+                    if (!this.tagsToMetricPointIndexDictionaryDelta.TryGetValue(givenTags, out lookupData))
+                    {
+                        if (this.reclaimMetricPoints)
+                        {
+                            // Check for an available MetricPoint
+                            if (this.availableMetricPoints!.Count > 0)
+                            {
+                                index = this.availableMetricPoints.Dequeue();
+                            }
+                            else
+                            {
+                                // No MetricPoint is available for reuse
+                                return -1;
+                            }
+                        }
+                        else
+                        {
+                            index = ++this.metricPointIndex;
+                            if (index == this.metricPointReclamationThreshold)
+                            {
+                                this.reclaimMetricPoints = true;
+                            }
+                        }
+
+                        lookupData = new LookupData(index, Tags.EmptyTags, givenTags);
+
+                        ref var metricPoint = ref this.metricPoints[index];
+                        metricPoint = new MetricPoint(this, this.aggType, givenTags.KeyValuePairs, this.histogramBounds, this.exponentialHistogramMaxSize, this.exponentialHistogramMaxScale, lookupData);
+
+                        // Add to dictionary *after* initializing MetricPoint
+                        // as other threads can start writing to the
+                        // MetricPoint, if dictionary entry found.
+
+                        // givenTags will always be sorted when tags length == 1
+                        this.tagsToMetricPointIndexDictionaryDelta.TryAdd(givenTags, lookupData);
+                    }
+                }
+            }
+        }
+
+        // Found the MetricPoint
+        index = lookupData.Index;
+        ref var metricPointAtIndex = ref this.metricPoints[index];
+        var referenceCount = Interlocked.Increment(ref metricPointAtIndex.ReferenceCount);
+
+        if (this.reclaimMetricPoints)
+        {
+            if (referenceCount < 0)
+            {
+                // Rare case: TakeSnapshot method had already marked the MetricPoint available for reuse as it has not been updated in last collect cycle.
+
+                index = this.RemoveStaleEntriesAndGetAvailableMetricPointRare(lookupData, length);
+            }
+            else
+            {
+                if (metricPointAtIndex.LookupData != lookupData)
```

Comment:
trying to understand when this will be true specifically.  

1) Thread 1 finds the tag at `573`
2) Snapshot collects the metricpoint at `323`
3) Thread 2 claims it `614`
4) Thread 1 now runs in to this?

Reply:
Consider this scenario:
- MetricPoint with the tags `k1,v1` gets stored at index 100 by some Update thread
- Snapshot collects the value for `k1,v1`
- No further updates are done to the `k1,v1` (Next Snapshot run would reclaim this MetricPoint)
- An update thread `T1` wants to update `k1,v1`, it retrieves the index as 100 and then gets switched out by CPU
- Snapshot method runs and reclaims index 100
- Another update thread `T2` uses index 100 to store a new MetricPoint `k2,v2`
- CPU now brings back update thread `T1` and it resumes with the index to be updated as 100 which is now tracking `k2,v2` and not `k1,v1`